spot-client/src/common/remote-control/P2PReconnectTrait.js

```diff
@@ -0,0 +1,160 @@
+
+        this._ensureP2PConnectionStarted();
+    }
+
+    /**
+     * Cancels future retry if scheduled.
+     *
+     * @private
+     * @returns {void}
+     */
+    _cancelRetry() {
+        clearTimeout(this._retryTimeout);
+        this._retryTimeout && logger.log(`Canceling P2P retry for: ${this._remoteAddress}`);
+        this._retryTimeout = undefined;
+    }
+
+    /**
+     * Deactivates the retry logic and cancels and scheduled retries.
+     *
+     * @returns {void}
+     */
+    deactivate() {
+        for (const unregister of this._unregisterListeners) {
+            unregister();
+        }
+        this._unregisterListeners = [];
+        this._cancelRetry();
+    }
+
+    /**
+     * Starts new P2P if there isn't any active one.
+     *
+     * @returns {boolean}
+     * @private
+     */
+    _ensureP2PConnectionStarted() {
+        if (!this._xmppConnection.isConnected()) {
+
+            return false;
+        }
+
+        const connection = this._p2pSignaling.getConnectionForAddress(this._remoteAddress);
+
+        if (connection && connection.isDataChannelActive()) {
+
+            return false;
+        } else if (connection && !connection.isDataChannelActive()) {
+            this._p2pSignaling.closeConnection(this._remoteAddress);
+        }
+
+        logger.log(`Starting P2P Connection with ${this._remoteAddress}`);
+
+        this._p2pSignaling.start(this._remoteAddress);
+
+        return true;
+    }
+
+    /**
+     * Triggered when P2P connection status changes.
+     *
+     * @param {string} remoteAddress - The remote address for the event target.
+     * @param {boolean} isReady - Whether the data channel is now active or not.
+     * @private
+     * @returns {void}
+     */
+    _onDataChannelReadyUpdate(remoteAddress, isReady) {
+        if (remoteAddress !== this._remoteAddress) {
+
+            return;
+        }
+
+        if (isReady) {
+            this._retry = 0;
+        } else {
+            if (!this._xmppConnection.isConnected()) {
+
+                return;
+            }
+            if (this._retry >= this._maxRetry) {
+                logger.log(`Giving up on P2P retries for ${this._remoteAddress}`);
+
+                return;
+            }
+
+            this._cancelRetry();
+
+            const retryDelay = P2PReconnectTrait.RETRY_DELAY;
+
+            logger.log(`Will retry P2P connection in ${retryDelay}ms`);
+
+            this._retryTimeout = setTimeout(() => {
+                if (this._ensureP2PConnectionStarted()) {
+                    this._retry += 1;
+                }
+                this._retryTimeout = undefined;
+            }, retryDelay);
+        }
+    }
+
+    /**
+     * Triggered whe signaling goes offline/online.
+     *
+     * @param {boolean} isConnected - The new signaling status.
+     * @returns {void}
+     */
+    _onSignalingConnectedUpdate(isConnected) {
+        isConnected && this._ensureP2PConnectionStarted();
+    }
+}
```

Comment:
I haven't heard of `trait` before. What does it mean?

Reply:
Well Im not english specialist so dont ask me lol. In this case it's kind of a feature enabled on another feature, which is doing reconnects on broken p2p session. Maybe it could be renamed to session. Was just trying to make it more about the reconnect.

Reply:
I think I took the word from scala: https://www.tutorialspoint.com/scala/scala_traits.htm it is used there as another way of calling interface, that has some implementation. Maybe this doesnt make sense. I can just call it reconnect decorator or something around those lines if we decide to make it a decorator?